src/torchcodec/samplers/_implem.py

```diff
@@ -0,0 +1,62 @@
+from typing import List
+
+import torch
+
+from torchcodec.decoders import (  # TODO: move FrameBatch to torchcodec.FrameBatch?
+    FrameBatch,
+    SimpleVideoDecoder,
+)
+
+
+def clips_at_random_indices(
+    decoder: SimpleVideoDecoder,
+    *,
+    num_clips: int = 1,
+    num_frames_per_clip: int = 1,
+    num_indices_between_frames: int = 1,
+) -> List[FrameBatch]:
+    if num_clips <= 0:
+        raise ValueError(f"num_clips ({num_clips}) must be strictly positive")
+    if num_frames_per_clip <= 0:
+        raise ValueError(
+            f"num_frames_per_clip ({num_frames_per_clip}) must be strictly positive"
+        )
+    if num_indices_between_frames <= 0:
+        raise ValueError(
+            f"num_indices_between_frames ({num_indices_between_frames}) must be strictly positive"
+        )
+
+    # Determine the span of a clip, i.e. the number of frames (or indices)
+    # between the first and last frame in the clip, both included. This isn't
+    # the same as the number of frames in a clip!
+    # Example: f means a frame in the clip, x means a frame excluded from the clip
+    # num_frames_per_clip = 4
+    # num_indices_between_frames = 1, clip = ffff      , span = 4
+    # num_indices_between_frames = 2, clip = fxfxfxf   , span = 7
+    # num_indices_between_frames = 3, clip = fxxfxxfxxf, span = 10
+    clip_span = num_indices_between_frames * (num_frames_per_clip - 1) + 1
+
+    # TODO: We should probably not error.
+    if clip_span > len(decoder):
+        raise ValueError(
+            f"Clip span ({clip_span}) is larger than the number of frames ({len(decoder)})"
+        )
```

Comment:
@ahmadsharif1 @scotts 

I am errorring here out of convenience, but I'm not sure this leads to the best UX. I think we should try as hard as we can to return something, even if it may not respect `num_indices_between_frames` or `num_frames_per_clip`.

Typically, users will apply consistent values for `num_indices_between_frames` and `num_frames_per_clip` across all videos. However, given the significant variability in video lengths, it's probable that any reasonable configuration for these parameters might not work for at least one video in a large dataset. And that would lead to a loud error mid-training, which can be problematic and frustrating.

In general what we want to avoid are silently-wrong result. I don't think there is a risk of silently wrong result here. E.g. if the output shape is *really* wrong, then the model will fail loudly later in the pipeline.

If `clip_span > len(decoder)`,  the strategy I would suggest is to:
- decrease `num_indices_between_frames` and try to return something that respects `num_frames_per_clip`
- if not, `reduce_num_frames_per_clip`

Of course, this should be properly documented.

Looking for your thoughts and input before I start implementing such a thing :)

Reply:
Agreed that this should not be an error condition for the reasons you describe.

The behavior that I think users will want might be annoying to implement. I think we have to respect `num_frames_per_clip`, even if it means we produce duplicate frames. Returning back clips that always respect `num_frames_per_clip` is, I think, one of the main values of the sampler. It prevents users from having to implement this annoying logic to make sure they have the right data shape for their model inputs. What I think that we should return:

* As many frames as possible that obey the users constraints. Let's call this number of frames `possible`.
* The remaining frames, `num_frames_per_clip - possible`, will be duplicates from within that clip. Users should be able to control this behavior with a policy. I can think of two reasonable policies to start with: `REPEAT_LAST_FRAME` and `REPEAT_WHOLE_CLIP`.

Reply:
The default should probably be to not error, but at the user's request (maybe as a parameter), there should be a strict mode where it errors out if there aren't enough frames to satisfy the requirements of the user.

Reply:
We chatted offline, we'll introduce a policy parameter and potentially another boolean flag, this is still TBD. Let me resolve this thread to clear up the long list of open comments.